Handle anonymous structs

# Problem

With a type which contains an array of anonymous struct:
```go
type uploadBatchBody struct {
    Data []struct {
        CreativeID string `json:"creative_id"`
    } `json:"data"`
}
```

We get a segfault: 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x58 pc=0x1185d82]
goroutine 1 [running]:
github.com/alexjomin/openapi-parser/docparser.parseNamedType(0xc00021dc00, 0x12c2800, 0xc000253a10, 0x0, 0x0, 0x0, 0x0)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/docparser/parser.go:106 +0x7e2
github.com/alexjomin/openapi-parser/docparser.(*openAPI).parseStructs(0xc00009f2b0, 0xc00021dc00, 0xc00026a3a0, 0x1b, 0xc00045d820)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/docparser/model.go:403 +0x5b0
github.com/alexjomin/openapi-parser/docparser.(*openAPI).parseSchemas(0xc00009f2b0, 0xc00021dc00)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/docparser/model.go:486 +0x648
github.com/alexjomin/openapi-parser/docparser.(*openAPI).Parse.func1(0xc0002af140, 0x33, 0x12c45c0, 0xc000268340, 0x0, 0x0, 0x18, 0xc000347600)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/docparser/model.go:254 +0xc6
path/filepath.walk(0xc0002af140, 0x33, 0x12c45c0, 0xc000268340, 0xc000347b10, 0x0, 0x0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:360 +0x425
path/filepath.walk(0xc0003f8ba0, 0x1a, 0x12c45c0, 0xc0002fcea0, 0xc000347b10, 0x0, 0x0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:384 +0x2ff
path/filepath.walk(0xc0003f8b40, 0x11, 0x12c45c0, 0xc0002fcdd0, 0xc000347b10, 0x0, 0x0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:384 +0x2ff
path/filepath.walk(0xc00023be80, 0xc, 0x12c45c0, 0xc0002f1e10, 0xc000347b10, 0x0, 0x0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:384 +0x2ff
path/filepath.walk(0xc00023be1a, 0x3, 0x12c45c0, 0xc0002f1ba0, 0xc000347b10, 0x0, 0x0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:384 +0x2ff
path/filepath.walk(0x126b364, 0x1, 0x12c45c0, 0xc00009f380, 0xc000347b10, 0x0, 0x16007d0)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:384 +0x2ff
path/filepath.Walk(0x126b364, 0x1, 0xc0000c5b10, 0x1258a20, 0xc0000c5b01)
    /usr/local/Cellar/go/1.14/libexec/src/path/filepath/path.go:406 +0xff
github.com/alexjomin/openapi-parser/docparser.(*openAPI).Parse(0xc00009f2b0, 0x126b364, 0x1, 0x1472ad8, 0x0, 0x0, 0x126bcfc, 0x6)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/docparser/model.go:260 +0x93
github.com/alexjomin/openapi-parser/cmd.glob..func2(0x1445c00, 0x1472ad8, 0x0, 0x0)
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/cmd/root.go:28 +0x19d
github.com/spf13/cobra.(*Command).execute(0x1445c00, 0xc00009a1c0, 0x0, 0x0, 0x1445c00, 0xc00009a1c0)
    /Users/user/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0x1445c00, 0x1040f4a, 0x142fd40, 0xc000000300)
    /Users/user/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
    /Users/user/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/alexjomin/openapi-parser/cmd.Execute()
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/cmd/root.go:40 +0x31
main.main()
    /Users/user/go/pkg/mod/github.com/alexjomin/openapi-parser@v1.3.0/main.go:6 +0x20
```

And if we just have the anonymous struct: 
```go
type uploadBatchBody struct {
    Data struct {
        CreativeID string `json:"creative_id"`
    } `json:"data"`
}
```

Get get a parsing error, but no crash:
```
ERRO[0000] Can't parse the type of field in struct       error="expr (&{%!s(token.Pos=369) %!s(*ast.FieldList=&{376 [0xc00016adc0] 436}) %!s(bool=false)}) not yet unsupported" field=Data
```

# Potential solution

An anonymous struct could be handled like so:

```go
// @openapi:schema
type SomeStruct struct {
    Data struct {
        ID string `json:"id"`
    } `json:"data"`
}
```

Outputs:
```yaml
SomeStruct:
    type: object
    properties:
        data:
            type: object
            properties:
                id:
                    type: string
```

--- a/docparser/datatest/user.go
+++ b/docparser/datatest/user.go
@@ -110,6 +110,17 @@ type Pet struct {
 	JSONData            json.RawMessage           `json:"json_data"`
 	CustomString        otherpackage.CustomString `json:"custom_string"`
 	Test                Test                      `json:"test"`
+	Anonymous           struct {
+		Field string `json:"field"`
+	} `json:"anonymous"`
+}
+
+// AnonymousArray struct
+// @openapi:schema
+type AnonymousArray struct {
+	Data []struct {
+		ID string `json:"id"`
+	} `json:"data"`
 }
 
 // Dog struct

--- a/docparser/datatest/user.go
+++ b/docparser/datatest/user.go
@@ -83,6 +83,7 @@ func PostFoo() {}
 // @openapi:schema
 type MapStringString map[string]string
 
+// WeirdInt type
 // @openapi:schema
 type WeirdInt int
 

--- a/docparser/model.go
+++ b/docparser/model.go
@@ -2,6 +2,7 @@ package docparser
 
 import (
 	"errors"
+	"fmt"
 	"go/ast"
 	"os"
 	"path/filepath"
@@ -290,7 +291,7 @@ func (spec *openAPI) Parse(path string, parseVendors []string, vendorsPath strin
 	spec.composeSpecSchemas()
 }
 
-func (spec *openAPI) parsePaths(f *ast.File) (errors []error) {
+func (spec *openAPI) parsePaths(f *ast.File) (errs []error) {
 	for _, s := range f.Comments {
 		t := s.Text()
 		// Test if comments is a path
@@ -310,7 +311,7 @@ func (spec *openAPI) parsePaths(f *ast.File) (errors []error) {
 				WithError(err).
 				WithField("content", content).
 				Error("Unable to unmarshal path")
-			errors = append(errors, &BuildError{
+			errs = append(errs, &BuildError{
 				Err:     err,
 				Content: content,
 				Message: "unable to unmarshal path",
@@ -328,6 +329,10 @@ func (spec *openAPI) parsePaths(f *ast.File) (errors []error) {
 							WithField("url", url).
 							WithField("verb", currentVerb).
 							Error("Verb for this path already exists")
+						errs = append(errs, &BuildError{
+							Err:     errors.New("verb for this path already exists"),
+							Content: fmt.Sprintf("url: %s, verb: %s", url, currentVerb),
+						})
 						continue
 					}
 					spec.Paths[url][currentVerb] = currentDesc
@@ -403,12 +408,21 @@ func (spec *openAPI) composeSpecSchemas() {
 	}
 }
 
-func (spec *openAPI) parseMaps(mp *ast.MapType) *schema {
+func (spec *openAPI) parseMaps(mp *ast.MapType) (*schema, []error) {
+	errors := make([]error, 0)
+
 	// only map[string]
 	if i, ok := mp.Key.(*ast.Ident); ok {
-		t, _, _ := parseIdentProperty(i)
+		t, _, err := parseIdentProperty(i)
+		if err != nil {
+			errors = append(errors, BuildError{
+				Err:     err,
+				Message: "could not parse ident property from map",
+			})
+		}
+
 		if t != "string" {
-			return nil
+			return nil, errors
 		}
 	}
 
@@ -418,13 +432,15 @@ func (spec *openAPI) parseMaps(mp *ast.MapType) *schema {
 
 	// map[string]interface{}
 	if _, ok := mp.Value.(*ast.InterfaceType); ok {
-		return &e
+		return &e, errors
 	}
 
-	return nil
+	return nil, errors
 }
 
-func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) interface{} {
+func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) (interface{}, []error) {
+	errors := make([]error, 0)
+
 	var cs *composedSchema
 	e := newEntity()
 	e.Type = "object"
@@ -443,6 +459,11 @@ func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) interface{} 
 
 			if err != nil {
 				logrus.WithError(err).WithField("field", fld.Names[0]).Error("Can't parse the type of field in struct")
+				errors = append(errors, BuildError{
+					Err:     err,
+					Content: fld.Names[0].String(),
+					Message: "can't parse the type of field in struct",
+				})
 				continue
 			}
 
@@ -465,6 +486,10 @@ func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) interface{} 
 			p, err := parseNamedType(f, fld.Type, nil)
 			if err != nil {
 				logrus.WithError(err).WithField("field", fld.Type).Error("Can't parse the type of composed field in struct")
+				errors = append(errors, BuildError{
+					Err:     err,
+					Message: "can't parse the type of composed field in struct",
+				})
 				continue
 			}
 
@@ -473,10 +498,10 @@ func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) interface{} 
 	}
 
 	if cs == nil {
-		return &e
+		return &e, errors
 	} else {
 		cs.AllOf = append(cs.AllOf, &e)
-		return cs
+		return cs, errors
 	}
 }
 
@@ -512,13 +537,23 @@ func (spec *openAPI) parseSchemas(f *ast.File) (errors []error) {
 
 				switch n := ts.Type.(type) {
 				case *ast.MapType:
-					entity = spec.parseMaps(n)
+					var errs []error
+					entity, errs = spec.parseMaps(n)
+					if len(errs) != 0 {
+						errors = append(errors, errs...)
+					}
+
 					logrus.
 						WithField("name", entityName).
 						Info("Parsing Schema")
 
 				case *ast.StructType:
-					entity = spec.parseStructs(f, n)
+					var errs []error
+					entity, errs = spec.parseStructs(f, n)
+					if len(errs) != 0 {
+						errors = append(errors, errs...)
+					}
+
 					mtd, ok := entity.(metaSchema)
 					if ok {
 						mtd.SetCustomName(entityName)
@@ -534,7 +569,6 @@ func (spec *openAPI) parseSchemas(f *ast.File) (errors []error) {
 						logrus.WithError(err).Error("Can't parse the type of field in struct")
 						errors = append(errors, &BuildError{
 							Err:     err,
-							Content: "",
 							Message: "Can't parse the type of field in struct",
 						})
 						continue
@@ -552,6 +586,10 @@ func (spec *openAPI) parseSchemas(f *ast.File) (errors []error) {
 					p, err := parseNamedType(f, ts.Type, nil)
 					if err != nil {
 						logrus.WithError(err).Error("can't parse custom type")
+						errors = append(errors, BuildError{
+							Err:     err,
+							Message: "can't parse custom type",
+						})
 						continue
 					}
 					p.SetCustomName(entityName)

--- a/docparser/model.go
+++ b/docparser/model.go
@@ -451,12 +451,11 @@ func (spec *openAPI) parseStructs(f *ast.File, tpe *ast.StructType) (interface{}
 			if j.ignore {
 				continue
 			}
-			p, err := parseNamedType(f, fld.Type, nil)
-
 			if j.required {
 				e.Required = append(e.Required, j.name)
 			}
 
+			p, err := parseNamedType(f, fld.Type, nil)
 			if err != nil {
 				logrus.WithError(err).WithField("field", fld.Names[0]).Error("Can't parse the type of field in struct")
 				errors = append(errors, BuildError{

--- a/docparser/parser.go
+++ b/docparser/parser.go
@@ -98,11 +98,18 @@ func parseNamedType(gofile *ast.File, expr ast.Expr, sel *ast.Ident) (*schema, e
 		p.Format = format
 		return &p, nil
 	case *ast.StarExpr: // pointer to something, optional by default
-		t, _ := parseNamedType(gofile, ftpe.X, nil)
+		t, err := parseNamedType(gofile, ftpe.X, nil)
+		if err != nil {
+			return nil, err
+		}
 		t.Nullable = true
 		return t, nil
 	case *ast.ArrayType: // slice type
-		cp, _ := parseNamedType(gofile, ftpe.Elt, nil)
+		cp, err := parseNamedType(gofile, ftpe.Elt, nil)
+		if err != nil {
+			return nil, err
+		}
+
 		if cp.Format == "binary" {
 			p.Type = "string"
 			p.Format = "binary"
@@ -123,7 +130,11 @@ func parseNamedType(gofile *ast.File, expr ast.Expr, sel *ast.Ident) (*schema, e
 	case *ast.StructType:
 		return nil, fmt.Errorf("expr (%s) not yet unsupported", expr)
 	case *ast.SelectorExpr:
-		t, _ := parseNamedType(gofile, ftpe.X, ftpe.Sel)
+		t, err := parseNamedType(gofile, ftpe.X, ftpe.Sel)
+		if err != nil {
+			return nil, err
+		}
+
 		return t, nil
 	case *ast.MapType:
 		k, kerr := parseNamedType(gofile, ftpe.Key, nil)

--- a/docparser/parser.go
+++ b/docparser/parser.go
@@ -98,14 +98,14 @@ func parseNamedType(gofile *ast.File, expr ast.Expr, sel *ast.Ident) (*schema, e
 		p.Format = format
 		return &p, nil
 	case *ast.StarExpr: // pointer to something, optional by default
-		t, err := parseNamedType(gofile, ftpe.X, nil)
+		t, err := parseNamedType(gofile, ftpe.X, sel)
 		if err != nil {
 			return nil, err
 		}
 		t.Nullable = true
 		return t, nil
 	case *ast.ArrayType: // slice type
-		cp, err := parseNamedType(gofile, ftpe.Elt, nil)
+		cp, err := parseNamedType(gofile, ftpe.Elt, sel)
 		if err != nil {
 			return nil, err
 		}
@@ -122,13 +122,34 @@ func parseNamedType(gofile *ast.File, expr ast.Expr, sel *ast.Ident) (*schema, e
 			if len(cp.Items) != 0 {
 				p.Items["items"] = cp.Items
 			}
+			if len(cp.Properties) != 0 {
+				p.Items["properties"] = cp.Properties
+			}
 		}
 		if cp.Ref != "" {
 			p.Items["$ref"] = cp.Ref
 		}
 		return &p, nil
 	case *ast.StructType:
-		return nil, fmt.Errorf("expr (%s) not yet unsupported", expr)
+		p = newEntity()
+		p.Type = "object"
+
+		for _, field := range ftpe.Fields.List {
+			j, err := parseJSONTag(ftpe.Fields.List[0])
+			if err != nil {
+				return nil, err
+			}
+
+			pnt, err := parseNamedType(gofile, field.Type, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			p.Properties[j.name] = pnt
+
+		}
+
+		return &p, nil
 	case *ast.SelectorExpr:
 		t, err := parseNamedType(gofile, ftpe.X, ftpe.Sel)
 		if err != nil {
@@ -137,8 +158,8 @@ func parseNamedType(gofile *ast.File, expr ast.Expr, sel *ast.Ident) (*schema, e
 
 		return t, nil
 	case *ast.MapType:
-		k, kerr := parseNamedType(gofile, ftpe.Key, nil)
-		v, verr := parseNamedType(gofile, ftpe.Value, nil)
+		k, kerr := parseNamedType(gofile, ftpe.Key, sel)
+		v, verr := parseNamedType(gofile, ftpe.Value, sel)
 		if kerr != nil || verr != nil || k.Type != "string" {
 			// keys can only be of type string
 			return nil, fmt.Errorf("expr (%s) not yet unsupported", expr)

--- a/docparser/parser_test.go
+++ b/docparser/parser_test.go
@@ -139,9 +139,51 @@ func TestParseNamedType(t *testing.T) {
 			}},
 		},
 		{
-			description:   "Should throw error when parse *ast.StructType",
-			expr:          &ast.StructType{},
-			expectedError: "expr (&{%!s(token.Pos=0) %!s(*ast.FieldList=<nil>) %!s(bool=false)}) not yet unsupported",
+			description: "Should *ast.ArrayType with *ast.StructType",
+			expr: &ast.ArrayType{
+				Elt: &ast.StructType{
+					Fields: &ast.FieldList{
+						List: []*ast.Field{
+							{
+								Type: &ast.Ident{Name: "string"},
+								Tag:  &ast.BasicLit{Value: "`json:\"str\"`"},
+							},
+						},
+					},
+				},
+			},
+			expectedSchema: &schema{
+				Type: "array",
+				Items: map[string]interface{}{
+					"type": "object",
+					"properties": map[string]*schema{
+						"str": {
+							Type: "string",
+						},
+					},
+				},
+			},
+		},
+		{
+			description: "Should *ast.StructType to anonymous struct",
+			expr: &ast.StructType{
+				Fields: &ast.FieldList{
+					List: []*ast.Field{
+						{
+							Type: &ast.Ident{Name: "string"},
+							Tag:  &ast.BasicLit{Value: "`json:\"str\"`"},
+						},
+					},
+				},
+			},
+			expectedSchema: &schema{
+				Type: "object",
+				Properties: map[string]*schema{
+					"str": {
+						Type: "string",
+					},
+				},
+			},
 		},
 		{
 			description: "Should throw error when parse *ast.MapType[nil]nil",

--- a/openapi.yaml
+++ b/openapi.yaml
@@ -41,6 +41,17 @@ paths:
       - petstore_auth:
         - write:pets
         - read:pets
+      servers:
+      - url: https://{environment}.hello.com
+        description: what up
+        variables:
+          environment:
+            default: api
+            enum:
+            - api
+            - api.dev
+            - api.staging
+            description: ""
     post:
       description: Returns all pets from the system that the user has access to
       responses:
@@ -62,6 +73,16 @@ paths:
               $ref: '#/components/schemas/Pet'
 components:
   schemas:
+    AnonymousArray:
+      type: object
+      properties:
+        data:
+          type: array
+          items:
+            properties:
+              id:
+                type: string
+            type: object
     AnyValue:
       description: 'Can be anything: string, number, array, object, etc., including
         `null`'
@@ -98,6 +119,11 @@ components:
           type: object
           additionalProperties:
             type: integer
+        anonymous:
+          type: object
+          properties:
+            field:
+              type: string
         children:
           type: object
           additionalProperties:
@@ -171,3 +197,4 @@ components:
           type: string
     WeirdInt:
       type: integer
+x-tagGroups: []

